chatterino2: fix install on macOS

strager opened the pull request:
On Darwin/macOS, chatterino2's install phase copies no files into the output. `make install` does not copy chatterino2.app. Copy the .app manually, fixing the installation.

Additionally, apply a workaround for issue #42893. Otherwise, the installed .app crashes unless qtbase is separately installed in the user's environment.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

strager committed:
chatterino2: fix install on macOS

On Darwin/macOS, chatterino2's install phase copies no files into the
output. `make install` does not copy chatterino2.app. Copy the .app
manually, fixing the installation.

With this patch, chatterino2 almost works on macOS. Issue #42893
applies, so to make chatterino2 work, you must install qtbase in your
environment. For example:

    $ nix-env -iA nixpkgs.qt5.qtbase
    $ open ~/.nix-profile/Applications/chatterino.app

Infinisil commented:
You should be able to use `wrapQtAppsHook` to not require the `nix-env` step, see https://nixos.org/nixpkgs/manual/#sec-language-qt and https://github.com/NixOS/nixpkgs/pull/54525

strager commented:
> You should be able to use `wrapQtAppsHook` to not require the `nix-env` step, see https://nixos.org/nixpkgs/manual/#sec-language-qt and #54525

Thanks for pointing this out to me! It looks like that feature has no effect on macOS binaries installed in Applications. I'll look into fixing that.

strager committed:
chatterino2: fix: Could not find the Qt platform plugin "cocoa" in ""

wrapQtAppsHook ignores Mach-O binaries [1]. Additionally, wrapQtAppsHook
doesn't look inside $out/Applications [2], which is where chatterino2 is
installed. This means chatterino2 for Darwin/macOS is not wrapped
properly, causing the following error when launched:

> qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
>
> This application failed to start because no Qt platform plugin could
> be initialized. Reinstalling the application may fix this problem.

On macOS, explicitly wrap chatterino2's executable.

[1] pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh line 85
[2] pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh line 76

Infinisil commented:
Ohh nice find, would be good if this was fixed in wrapQtAppsHook directly

worldofpeace commented on pkgs/applications/networking/instant-messengers/chatterino2/default.nix:
```diff
@@ -10,8 +10,15 @@ mkDerivation rec {
     fetchSubmodules = true;
   };
-  nativeBuildInputs = [ qmake pkgconfig ];
+  nativeBuildInputs = [ qmake pkgconfig wrapQtAppsHook ];
```
this already uses the custom deriver that add `wrapQtAppsHook`, so you don't need to add it here explicitly.

strager replied:
I understand that the dependency is implied. I think it's clearer if the dependency is explicit, though.

worldofpeace replied:
It's not implied, the deriver adds the dependency by design, and in 19.09 we had fix several 100+ expressions to use `mkDerivation` instead of `stdenv.mkDerivation` because people didn't read the docs.
It was even requested in the PR that initially was meant fix this https://github.com/NixOS/nixpkgs/pull/78296 https://github.com/NixOS/nixpkgs/compare/c2ce79000873d1936c07ec0261c200927c2bb40c..717c2207c2a0fe702ff678af230e583e05deb177

strager replied:
> It's not implied, the deriver adds the dependency by design

That's what I meant when I said "the dependency is implied". The dependency is implied when using mkDerivation from pkgs/development/libraries/qt-5/mkDerivation.nix (instead of stdenv.mkDerivation).

On line 20, we call wrapQtApp in postFixup (on Darwin). wrapQtApp requires a dependency on wrapQtAppsHook. That's why I added the dependency here on line 13. If expressing the dependency here is a bug, let me know so I can remove it. My current understanding is that expressing the dependency here is merely *optional*.

strager commented:
ping

cc @rexim, @doronbehar

doronbehar commented:
Can't test on darwin @strager Sorry :/.
